R/env-special.R

```diff
@@ -197,44 +197,51 @@ current_env <- function() {
 #' These functions are experimental and may not belong to the rlang
 #' package. Expect API changes.
 #'
-#' @param pkg The name of a package. If `NULL`, the surrounding
-#'   namespace is returned, or an error is issued if not called within
-#'   a namespace. If a function, the enclosure of that function is
-#'   checked.
+#' @param pkg The name of a package or an environment. If a function,
+#'   the environment of that function is searched for a namespace with
+#'   [base::topenv()].
 #'
 #' @seealso [pkg_env()]
 #' @keywords internal
 #' @export
 ns_env <- function(pkg = NULL) {
-  if (is_null(pkg)) {
-    bottom <- topenv(caller_env())
-    if (!isNamespace(bottom)) abort("not in a namespace")
-    bottom
-  } else if (is_function(pkg)) {
-    env <- env_parent(pkg)
-    if (isNamespace(env)) {
-      env
-    } else {
-      NULL
-    }
-  } else {
-    asNamespace(pkg)
+  env <- switch(typeof(pkg),
+    builtin = ,
+    special =
```

Comment:
It looks weird to indent the switch in this way — I'd do `character = if (is_string(pkg)) asNamespace(pkg)` so that every case can fit on one line.

Reply:
I like this indent because it's a little more scalable, e.g. you can add `{ }` braces when a couple of expressions are needed and it's not worth moving them to a function.

Your style works here though.